python/bind_operators.h

```diff
@@ -34,21 +37,26 @@ void bind_common_operators(pybind11::class_<T, Ignored...> &c) {
       "__copy__", [](T &self) { return self; },
       py::call_guard<py::gil_scoped_release>(), "Return a (shallow) copy.");
   c.def(
-      "__deepcopy__", [](T &self, py::dict) { return copy(self); },
+      "__deepcopy__", [](T &self, const py::dict &) { return copy(self); },
       py::call_guard<py::gil_scoped_release>(), "Return a (deep) copy.");
 }
 
 template <class T, class... Ignored>
 void bind_astype(py::class_<T, Ignored...> &c) {
   c.def(
       "astype",
-      [](const T &self, const scipp::DType type, const bool copy) {
-        return astype(self, type,
-                      copy ? scipp::CopyPolicy::Always
-                           : scipp::CopyPolicy::TryAvoid);
+      [](const T &self, const py::object &type, const bool copy) {
+        const auto [scipp_dtype, dtype_unit] =
+            cast_dtype_and_unit(type, std::nullopt);
+        const auto unit =
+            dtype_unit == scipp::units::one ? self.unit() : dtype_unit;
+        [[maybe_unused]] py::gil_scoped_release release;
+        return to_unit(astype(self, scipp_dtype,
+                              copy ? scipp::CopyPolicy::Always
+                                   : scipp::CopyPolicy::TryAvoid),
+                       unit, scipp::CopyPolicy::TryAvoid);
```

Comment:
Is this for datetime? I think we should *not* support this. Unlike numpy, scipp tries to separate dtype and unit, and I think we should do the same here.